File: 1c_xml/python/minidom.py
import xml.dom.minidom as md
import io

if __name__ == "__main__":
    dom = md.parse('../cd-list.xml')
    for child in dom.childNodes[0].childNodes:
        print('childnode', child)

    for title in dom.getElementsByTagName('title'):
        print('title', title.firstChild.nodeValue)

    for track in dom.getElementsByTagName('track'):
        print('track-no', track.getAttribute('no'))

    root = dom.childNodes[0]
    newcd = dom.createElement('cd')
    artist = dom.createElement('artist')
    artist.appendChild(dom.createTextNode('Bob'))
    artist.setAttribute('country', 'DE')
    newcd.appendChild(artist)
    title = dom.createElement('title')
    title.appendChild(dom.createTextNode('Song for Alice'))
    newcd.appendChild(title)
    root.appendChild(newcd)
    fd = io.StringIO()
    dom.writexml(fd)
    print(fd.getvalue())

File: 1c_xml/rust/xml/src/bin/minidom.rs
use minidom::Element;

const NS: &str = "";
fn main() -> Result<(), Box<dyn std::error::Error + 'static>> {
    println!("minidom");
    let data = std::fs::read_to_string("../../cd-list.xml")?;
    let mut root: Element = data.parse().unwrap();

    for child in root.children() {
        println!("{:?}", child);
    }

    for child in root.children().filter(|e| e.is("title", NS)) {
        println!("title {:?}", child.text());
    }
    for child in root.children().filter(|e| e.is("cd", NS)) {
        for tl in child.children().filter(|e| e.is("track-list", NS)) {
            for track in tl.children().filter(|e| e.is("track", NS)) {
                println!("track {:?}", track.attr("no"));
            }
        }
    }

    let newcd = Element::builder("cd", "")
        .append(
            Element::builder("artist", "")
                .append("bob")
                .attr("country", "DE"),
        )
        .append(Element::builder("title", "").append("Song for Alice"))
        .build();
    root.append_child(newcd);
    let mut buf = Vec::<u8>::new();
    root.write_to(&mut buf).unwrap();
    println!("{}", String::from_utf8(buf).unwrap());
    Ok(())
}
